File: sipstack-transaction/src/main/scala/io/sipstack/transaction/InviteServerTransaction.scala
/**
 *
 */
package io.sipstack.transaction

import akka.actor.Actor
import io.sipstack.transport.FlowActor.IncomingRequest
import io.pkts.packet.sip.SipResponse
import akka.actor.ActorRef
import io.sipstack.transport.FlowActor.IncomingRequest
import io.sipstack.config.TransactionLayerConfiguration
import io.sipstack.transport.FlowActor.OutgoingResponse
import io.sipstack.transport.FlowActor.OutgoingResponse
import akka.actor.PoisonPill

/**
 * @author jonas@jonasborjesson.com
 *
 */
object InviteServerTransaction {

  
}

/**
 * Implements the Invite Server Transaction as specified by
 * rfc3261 section
 * 
 * <pre>
 *                     |INVITE
 *                     |pass INV to TU
 *  INVITE             V send 100 if TU won't in 200ms
 *  send response+-----------+
 *      +--------|           |--------+101-199 from TU
 *      |        | Proceeding|        |send response
 *      +------->|           |<-------+
 *               |           |          Transport Err.
 *               |           |          Inform TU
 *               |           |--------------->+
 *               +-----------+                |
 *  300-699 from TU |     |2xx from TU        |
 *  send response   |     |send response      |
 *                  |     +------------------>+
 *                  |                         |
 *  INVITE          V          Timer G fires  |
 *  send response+-----------+ send response  |
 *      +--------|           |--------+       |
 *      |        | Completed |        |       |
 *      +------->|           |<-------+       |
 *               +-----------+                |
 *                  |     |                   |
 *              ACK |     |                   |
 *              -   |     +------------------>+
 *                  |        Timer H fires    |
 *                  V        or Transport Err.|
 *               +-----------+  Inform TU     |
 *               |           |                |
 *               | Confirmed |                |
 *               |           |                |
 *               +-----------+                |
 *                     |                      |
 *                     |Timer I fires         |
 *                     |-                     |
 *                     |                      |
 *                     V                      |
 *               +-----------+                |
 *               |           |                |
 *               | Terminated|<---------------+
 *               |           |
 *               +-----------+
 * 
 * </pre>
 */
class InviteServerTransaction(flow:ActorRef, next:ActorRef, config:TransactionLayerConfiguration, invite:IncomingRequest) extends Actor {
  
  import scala.concurrent.duration._
  import context._
  
  /**
   * Whenever we receives a re-transmission we will re-send the last
   * response.
   */
  var lastResponse:SipResponse = null
  
  override def receive = init
  
  override def preStart {
    // val resp100Trying = req.request.createResponse(100);
    // system.scheduler.scheduleOnce(200 millis, self, resp100Trying)
  }
  
  def init:Receive = {
    case `invite` => processInitialInvite
    case _ => println("Strange, shouldn't happen or should it?")
  }
  
  /**
   * The very first message MUST be the same one that this [[InviteServerTransaction]]
   * was constructed with and once it has been received, we will do some initialization
   * and then transition to the proceeding state.
   */
  private def processInitialInvite:Unit = {
    system.scheduler.scheduleOnce(200 millis, self, "100trying")
    become(proceeding)
    relay(invite)
  }
  
  /**
   * Implements the proceeding state where we will
   * 1. retransmit the latest response if we receive a retransmitted INVITE
   * 2. any response received by the TU will be sent out.
   * 2.1. If reeponse is 101-199 -> stay in current state
   * 2.2 if response is 2xx -> transition to terminated state
   * 2.3 if response is error -> transition to completed state
   */
  def proceeding:Receive = {
    case event:IncomingRequest => println("received a request, better be a retransmitted invite " + event.msg)
    case event:OutgoingResponse => {
      relay(event)
      if (event.msg.isSuccess()) {
        terminate
      } else if (event.msg.isFinal()) {
        println("Received error response, transition to completed");
        // schedule timers etc
        become(completed)
      }
    }
    case "100trying" => {
        if (lastResponse == null) {
            val now = System.currentTimeMillis()
            val trying = invite.msg createResponse 100
            relay(new OutgoingResponse(now, invite.connectionId, invite.transactionId, invite.callId, trying))
        }
    }
    case unknown => println("Received some unknown shit while in proceeding" + unknown); 
  }
  
  private def terminate {
    become(terminated)
    self ! PoisonPill
  }
  
  def completed:Receive = {
    case event:IncomingRequest if event.msg.isAck => {
      become(confirmed)
    }
    
    case unknown => println("Received some unknown shit while in completed" + unknown); 
  }
  
  def terminated:Receive = {
    case unknown => println("ok, got something in the terminated state " + unknown)
  }
  
  def confirmed:Receive = {
    ???
  }
  
  private def relay(response:OutgoingResponse) {
    if (lastResponse == null || lastResponse.getStatus < response.msg.getStatus) {
      lastResponse = response.msg
    }
    flow ! response.msg
  }
  
  private def relay(msg:IncomingRequest): Unit = {
    next ! msg
  }
  
  
  
}

File: sipstack-netty-codec-sip/src/main/java/io/sipstack/netty/codec/sip/transaction/InviteServerTransaction.java
package io.sipstack.netty.codec.sip.transaction;

import io.pkts.packet.sip.SipMessage;
import io.pkts.packet.sip.SipRequest;
import io.pkts.packet.sip.SipResponse;
import io.sipstack.netty.codec.sip.actor.ActorSupport;
import io.sipstack.netty.codec.sip.actor.Cancellable;
import io.sipstack.netty.codec.sip.config.TimersConfiguration;
import io.sipstack.netty.codec.sip.config.TransactionLayerConfiguration;
import io.sipstack.netty.codec.sip.event.Event;
import io.sipstack.netty.codec.sip.event.SipMessageEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Consumer;

/**
 * Implements the Invite Server Transaction state machine as specified by rfc6026.
 *
 * <pre>
 *
 *                                    |INVITE
 *                                    |pass INV to TU
 *                 INVITE             V send 100 if TU won't in 200 ms
 *                 send response+------------+
 *                     +--------|            |--------+ 101-199 from TU
 *                     |        |            |        | send response
 *                     +------->|            |<-------+
 *                              | Proceeding |
 *                              |            |--------+ Transport Err.
 *                              |            |        | Inform TU
 *                              |            |<-------+
 *                              +------------+
 *                 300-699 from TU |    |2xx from TU
 *                 send response   |    |send response
 *                  +--------------+    +------------+
 *                  |                                |
 * INVITE           V          Timer G fires         |
 * send response +-----------+ send response         |
 *      +--------|           |--------+              |
 *      |        |           |        |              |
 *      +------->| Completed |<-------+      INVITE  |  Transport Err.
 *               |           |               -       |  Inform TU
 *      +--------|           |----+          +-----+ |  +---+
 *      |        +-----------+    | ACK      |     | v  |   v
 *      |          ^   |          | -        |  +------------+
 *      |          |   |          |          |  |            |---+ ACK
 *      +----------+   |          |          +->|  Accepted  |   | to TU
 *      Transport Err. |          |             |            |<--+
 *      Inform TU      |          V             +------------+
 *                     |      +-----------+        |  ^     |
 *                     |      |           |        |  |     |
 *                     |      | Confirmed |        |  +-----+
 *                     |      |           |        |  2xx from TU
 *       Timer H fires |      +-----------+        |  send response
 *       -             |          |                |
 *                     |          | Timer I fires  |
 *                     |          | -              | Timer L fires
 *                     |          V                | -
 *                     |        +------------+     |
 *                     |        |            |<----+
 *                     +------->| Terminated |
 *                              |            |
 *                              +------------+
 *
 * </pre>
 *
 * @author jonas@jonasborjesson.com
 */
public class InviteServerTransaction extends ActorSupport<Event, TransactionState> implements Transaction {

    private static final Logger logger = LoggerFactory.getLogger(InviteServerTransaction.class);

    private final TransactionId id;

    private final SipRequest originalInvite;

    private final TransactionLayerConfiguration config;

    private final TimersConfiguration timersConfig;

    private SipResponse lastResponse;

    private Cancellable timer100Trying;

    /**
     * If we enter the completed state we will setup Timer G for response retransmissions if this
     * transaction is over an unreliable transport.
     */
    private Cancellable timerG;


    /**
     * How many times Timer G has fired.
     */
    private int timerGCount;

    /**
     * If we enter the completed state, Timer H will take us from completed to terminated unless we
     * do receive an ACK before then. Hence, Timer H determines the longest amount of time we can
     * stay in the completed state.
     */
    private Cancellable timerH;

    /**
     * If we enter the accepted state, Timer L will let us know when it is time to transition
     * over to the terminated state. The accepted state is a "new" state introduced by RFC6026
     * in order to deal with re-transmitted INVITEs. In the old state machine, we transitioned
     * directly to terminated state as soon as we got a 2xx to the INVITE which meant that
     * any re-transmissions on the INVITE created a brand new transaction, which is not what
     * we want.
     */
    private Cancellable timerL;


    /**
     * @param id
     * @param initialState
     * @param values
     */
    protected InviteServerTransaction(final TransactionId id,
                                      final SipRequest invite,
                                      final TransactionLayerConfiguration config) {
        super(id.toString(), TransactionState.INIT, TransactionState.values());
        this.id = id;
        this.originalInvite = invite;
        this.config = config;
        this.timersConfig = config.getTimers();

        when(TransactionState.INIT, init);

        // when(TransactionState.PROCEEDING, proceeding);
        // onEnter(TransactionState.PROCEEDING, onEnterProceeding);
        // onExit(TransactionState.PROCEEDING, onExitProceeding);
    }

    /**
     * The init state. Just make sure that the first event we receive is the same INVITE as created
     * the transaction (yes, we compare references in this case, that's what we want) and then
     * transition over to the proceeding state.
     */
    private final Consumer<Event> init = event -> {
        if (event.isSipMessageEvent() && event.toSipMessageEvent().message() == originalInvite()) {
            ctx().forwardUpstream(event);
        } else {
            System.err.println("Queue??? shouldn't be able to happen. " + event);
        }
        become(TransactionState.PROCEEDING);
    };

    /**
     * Implements the proceeding state as follows:
     *
     * <pre>
     *
     *                                    |INVITE
     *                                    |pass INV to TU
     *                 INVITE             V send 100 if TU won't in 200 ms
     *                 send response+------------+
     *                     +--------|            |--------+ 101-199 from TU
     *                     |        |            |        | send response
     *                     +------->|            |<-------+
     *                              | Proceeding |
     *                              |            |--------+ Transport Err.
     *                              |            |        | Inform TU
     *                              |            |<-------+
     *                              +------------+
     *                 300-699 from TU |    |2xx from TU
     *                 send response   |    |send response
     *                  +--------------+    +------------+
     *                  |                                |
     *                  V                                V
     *         +------------+                     +------------+
     *         |  Completed |                     |  Accepted  |
     *         +------------+                     +------------+
     * </pre>
     */
    private final Consumer<Event> proceeding = event -> {
        if (event.isSipMessageEvent()) {
            final SipMessageEvent sipMsgEvent = event.toSipMessageEvent();
            final SipMessage msg = sipMsgEvent.message();

            if (isRetransmittedInvite(msg)) {
                if (lastResponse != null) {
                    relayResponse(event.toSipMessageEvent());
                }
                return;
            }

            final SipResponse response = msg.toResponse();
            relayResponse(sipMsgEvent);
            if (response.isSuccess()) {
                become(TransactionState.ACCEPTED);
                // become(TransactionState.TERMINATED);
            } else if (response.isFinal()) {
                become(TransactionState.COMPLETED);
            }
        } else if (event.isSipTimer100Trying()) {
            final SipResponse response = originalInvite().createResponse(100);
            ctx().forwardDownstream(null);
        } else {
            unhandled(event);
        }
    };

    /**
     * When entering the Proceeding state we may send a 100 Trying right away (depending on configuration) or we may
     * delay it with 200 ms and send it later unless TU already have sent some kind of response (any kind really)
     */
    private final Consumer<Event> onEnterProceeding = event -> {
        if (config().isSend100TryingImmediately()) {
            final SipMessage invite = originalInvite();
            if (invite.isAck()) {
                System.err.println("What the fuck, this is an ACK!");
            }
            final SipResponse response = invite.createResponse(100);
            // downstream.tell(IOWriteEvent.create(response), self());
        } else {
            // TODO: schedule timer...
            // timer100Trying = ctx().scheduler().schedule();
        }
    };

    private final Consumer<Event> onExitProceeding = event -> {
        if (timer100Trying != null) {
            timer100Trying.cancel();
        }
    };

    private void relayResponse(final SipMessageEvent event) {
        if (lastResponse == null
                || lastResponse.getStatus() < event.message().toResponse().getStatus()) {
            lastResponse = event.message().toResponse();
        }
        ctx().forwardDownstream(event);
    }

    /**
     * Check whether the request is a re-transmitted INVITE.
     * @param msg
     * @return
     */
    private boolean isRetransmittedInvite(SipMessage msg) {
        if (msg.isRequest() && msg.isInvite()) {
            // TODO: actually check it...
            return true;
        }
        return false;
    }

    private TransactionLayerConfiguration config() {
        return config;
    }

    private SipRequest originalInvite() {
        return originalInvite;
    }

    @Override
    protected final Logger logger() {
        return logger;
    }
}
